ranker/shared/ranker_obj.py

```diff
@@ -1,131 +1,167 @@
+        self.kedge_by_id = {n: self.kgraph["edges"][n] for n in kedges}
         self.kedges_by_knodes = defaultdict(list)
 
         for e in kedges:
-            self.kedges_by_knodes[tuple(sorted([kedges[e]['subject'], kedges[e]['object']]))].append(kedges[e])
+            self.kedges_by_knodes[
+                tuple(sorted([kedges[e]["subject"], kedges[e]["object"]]))
+            ].append(kedges[e])
 
         # find leaf-set qnodes
         degree = defaultdict(int)
-        for edge in self.qgraph['edges']:
-            degree[self.qgraph['edges'][edge]['subject']] += 1
-            degree[self.qgraph['edges'][edge]['object']] += 1
+        for edge in self.qgraph["edges"]:
+            degree[self.qgraph["edges"][edge]["subject"]] += 1
+            degree[self.qgraph["edges"][edge]["object"]] += 1
         self.leaf_sets = [
             node
-            for node in self.qgraph['nodes']
-            if self.qgraph['nodes'][node].get('is_set', False) and degree[node] == 1
+            for node in self.qgraph["nodes"]
+            if self.qgraph["nodes"][node].get("is_set", False) and degree[node] == 1
         ]
 
-
     def rank(self, answers, jaccard_like=False):
         """Generate a sorted list and scores for a set of subgraphs."""
         # get subgraph statistics
-        #print(f'{len(answers)} answers')
         answers_ = []
         scores_for_sort = []
         for answer in answers:
             answers_.append(self.score(answer, jaccard_like=jaccard_like))
-            scores_for_sort.append(max(analysis['score'] for analysis in answer.get('analyses',[{'score':0}])))
-        answers = [x for _, x in sorted(zip(scores_for_sort, answers), key=lambda pair: pair[0])]
+            scores_for_sort.append(
+                max(
+                    analysis["score"]
+                    for analysis in answer.get("analyses", [{"score": 0}])
+                )
+            )
+        answers = [
+            x
+            for _, x in sorted(zip(scores_for_sort, answers), key=lambda pair: pair[0])
+        ]
         return answers
 
     def score(self, answer, jaccard_like=False):
         """Compute answer score."""
         # answer is a list of dicts with fields 'id' and 'bound'
         r_node_ids, edges_all = self.get_rgraph(answer)
-        
-        for i_analysis,edges in enumerate(edges_all):
-            laplacian = self.graph_laplacian((r_node_ids[i_analysis],edges), answer)
-            if np.any(np.all(np.abs(laplacian) == 0, axis=0)):
-                answer['analyses'][i_analysis]['score'] = 0
-                continue
-            index = {node_id[0]: r_node_ids[i_analysis].index(node_id) for node_id in r_node_ids[i_analysis]}
-            q_node_ids = list(self.qgraph['nodes'].keys())
+
+        for i_analysis, edges in enumerate(edges_all):
+            laplacian = self.graph_laplacian((r_node_ids[i_analysis], edges), answer)
```

Comment:
Could this be moved down to where you remove the extra nodes? (just for easier code reading)